virtual_rainforest/abiotic/radiation.py

```diff
@@ -0,0 +1,231 @@
+            sunshine_hours: NDArray[np.float32], fraction of sunshine hours
+            albedo_vis: NDArray[np.float32], visible light albedo, default = 0.03
+
+        Returns:
+            tau: NDArray[np.float32], transmissivity, unitless
+            self.ppfd: NDArray[np.float32], photosynthetic photon flux density [mol m-2]
+
+        Reference:
+            Davis et al. (2017): Simple process-led algorithms for simulating habitats
+            (SPLASH v.1.0): robust indices of radiation, evapotranspiration and plant-
+            available moisture, Geosci. Model Dev., 10, 689-708
+        """
+        # calculate transmissivity (tau), unitless
+        tau_o = CLOUDY_TRANSMISSIVITY + TRANSMISSIVITY_COEFFICIENT * sunshine_hours
+        tau = tau_o * (1.0 + BEER_REGRESSION * self.elevation)
+
+        # Calculate daily photosynth. photon flux density (ppfd_d), mol/m^2
+        self.ppfd_d = (
+            (1.0e-6) * FLUX_TO_ENERGY * (1.0 - albedo_vis) * tau * shortwave_in
+        )
+        return tau
+
+    def calc_topofcanopy_radiation(
+        self,
+        tau: NDArray[np.float32],
+        shortwave_in: NDArray[np.float32],
+        sunshine_hours: NDArray[np.float32],
+        albedo_shortwave: NDArray[np.float32] = np.array(0.17, dtype=float),
+    ) -> None:
+        """Calculate top of canopy shortwave radiation.
+
+        Args:
+            shortwave_in: NDArray[np.float32], daily downward shortwave radiation[W m-2]
+            sunshine_hours: NDArray[np.float32], fraction of sunshine hours
+            albedo_shortwave: NDArray[np.float32], shortwave albedo, default = 0.17
+
+        Returns:
+            self.topofcanopy_radiation: NDArray[np.float32], shortwave radiation [W m-2]
+        """
+        self.topofcanopy_radiation = (1.0 - albedo_shortwave) * tau * shortwave_in
+
+    def calc_longwave_radiation(
+        self,
+        canopy_temperature: NDArray[np.float32],
+        surface_temperature: NDArray[np.float32],
+    ) -> tuple[NDArray[np.float32], NDArray[np.float32]]:
+        """Calculates longwave emission from canopy and forest floor.
+
+        Args:
+            canopy_temperature: NDArray[np.float32], canopy temperature of n layers [C]
+            surface_temperature: NDArray[np.float32], surface soil temperature [C]
+
+        Returns:
+            longwave_canopy: NDArray[np.float32], longwave radiation from n individual
+                canopy layers [W m-2]
+            longwave_soil: NDArray[np.float32], longwave radiation from soil [W m-2]
+            self.longwave_out: NDArray[np.float32], total longwave radiation [W m-2]
+        """
+        raise NotImplementedError("Implementation of this feature is missing")
+
+        # longwave emission canopy
+        # TODO define the longwave_canopy array to match grid+vertical structure
+        # which is not uniform (different number of layers in each grid cell)
+        longwave_canopy = np.full_like(canopy_temperature, 0)
```

Comment:
I already know that this needs changing, need to dive into core docs, so please skip

Reply:
I think this is going to be _much_ easier if we assert a configurable maximum number of canopy layers. That would give us a `canopy_temperature` array that would have the shape (n_cells, n_canopy_layers) and we could then do:

```python
longwave_canopy = (CANOPY_EMISSIVITY * BOLZMAN_CONSTANT * canopy_temperature ** 4)
```

There are three issues:

 - the memory use is higher
 - we need to keep track of which of those layers are "above" the canopy
 - we either have to abort the simulation if we exceed that number of layers (which might be good if the number of layers is not sane) or alternatively have some kind of mechanism for extending that maximum and adding layers to the canopy as needed.

The alternative here is dealing with ragged arrays, which neither `numpy` nor `xarray` handle elegantly. There are some really neat tricks for speeding up calculations on ragged arrays but it seems best to use memory rather than needing that complexity. @alexdewar @dalonsoa ?

Reply:
The canopy model is then basically going to populate an array `canopy_layer_heights` with shape (n_cells, n_canopy_layers, n_timesteps) and for any one timestep you might have in an example with two cells and five canopy layers:

```python
In [22]: np.array([[16.1, 19.6, 23.1, np.nan, np.nan],[10.1, 15.9, 20.1, 23.5, np.nan]])
Out[22]: 
array([[16.1, 19.6, 23.1,  nan,  nan],
       [10.1, 15.9, 20.1, 23.5,  nan]])
```

Reply:
Thanks David, I'll consider that as we discussed in person today